File: DS-Solutions/Python_Solutions/Twopointers_1.py
arr=[1,2,3,4,5,5,5,6,6,7,11]
i,j=0,len(arr)-1
n=11
ans=0
while i<j:
    if (arr[i]+arr[j])<n:
        i+=1
    elif (arr[i]+arr[j])>n:
        j-=1
    else:
        x = arr[i]
        xx = i
        while (i < j and arr[i] == x):
            i += 1
        y = arr[j]
        yy = j
        while (j >= i and arr[j] == y):
            j -= 1
        if (x == y):
            temp = i - xx + yy - j - 1
            ans += (temp * (temp + 1)) // 2
        else:
            ans += (i - xx) * (yy - j)
print(ans)

File: DS-Solutions/Cpp_Solutions/TwoPointers_1.cpp
//Given an array arr[] and an integer n, the task is to find the number of pairs of integers in the array whose sum is equal to n.
#include<bits/stdc++.h>
using namespace std;

int main()
{
	int arr[]={1,5,5,5,6,6,11};
	int size=sizeof arr/sizeof arr[0];
	int x=0,y=size-1;
	int n=10;
	int ans=0;
	while(x<y)
	{
		if(arr[x]+arr[y]<n)
		{
			x++;
		}
		else if(arr[x]+arr[y]>n)
		{
			y--;
		}
		else
		{
			int value1=arr[x],initial_pos1=x;
			while(x<y and arr[x]==value1)
				x++;
			int value2=arr[y],initial_pos2=y;
			while(y>=x and arr[y]==value2)
				y--;
			if(value1==value2)                      
			{
				int temp=x-initial_pos1+initial_pos2-y;
				ans+=(temp*(temp-1))/2;            
			}
			else
			{
				ans+=(x-initial_pos1)*(initial_pos2-y);
			}
		}
	}
	cout<<ans;
}
